# Data Cleaning with Pandas

In [2]:
import pandas as pd
import numpy as np

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it with `brew install wget`, or use `curl <url> -o <filename>`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the file that you want to unzip.

In [4]:
sales_df = pd.read_csv('data/Real Property Sales.zip')

/Users/anthonyschams/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [5]:
sales_df.head()

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

In [6]:
sales_df.describe()

ExciseTaxNbr     SalePrice  PropertyType  PrincipalUse  SaleInstrument  \
count  2.026164e+06  2.026164e+06  2.026164e+06  2.026164e+06    2.026164e+06   
mean   2.033068e+06  5.908315e+05  3.146618e+00  4.604457e+00    7.351473e+00   
std    5.633690e+05  5.994676e+06  3.831775e+00  2.619647e+00    6.613777e+00   
min    4.565830e+05 -6.000000e+02  0.000000e+00  0.000000e+00    0.000000e+00   
25%    1.576158e+06  0.000000e+00  3.000000e+00  2.000000e+00    3.000000e+00   
50%    2.044634e+06  1.500000e+05  3.000000e+00  6.000000e+00    3.000000e+00   
75%    2.511022e+06  3.495000e+05  3.000000e+00  6.000000e+00    1.500000e+01   
max    2.990860e+06  7.398850e+08  9.900000e+01  1.100000e+01    2.800000e+01   

         SaleReason  PropertyClass  
count  2.026164e+06   2.026164e+06  
mean   5.410099e+00   6.508585e+00  
std    6.238406e+00   2.484930e+00  
min    0.000000e+00   0.000000e+00  
25%    1.000000e+00   3.000000e+00  
50%    1.000000e+00   8.000000e+00  
75%    1.000000e+01   8.000000e+00  
max    1.900000e+01   9.000000e+00

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [29]:
small_sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']].copy()

In [8]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.8+ MB


In [9]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/Users/anthonyschams/anaconda/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Another warning! Which column has index 11?

In [10]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [11]:
bldg_df.head().T

0                           1  \
Major                                   62304                       62304   
Minor                                    9347                        9378   
BldgNbr                                     1                           1   
NbrLivingUnits                              1                           1   
Address              907  SW 104TH ST   98146  10215   11TH AVE SW  98146   
BuildingNumber                          907                         10215   
Fraction                                                                    
DirectionPrefix                            SW                               
StreetName          104TH                       11TH                        
StreetType                             ST                          AVE      
DirectionSuffix                                                        SW   
ZipCode                                 98146                       98146   
Stories                                     1                           1   
BldgGrade                                   7                           7   
BldgGradeVar                                0                           0   
SqFt1stFloor                             1180                         970   
SqFtHalfFloor                               0                           0   
SqFt2ndFloor                                0                           0   
SqFtUpperFloor                              0                           0   
SqFtUnfinFull                               0                           0   
SqFtUnfinHalf                               0                           0   
SqFtTotLiving                            1570                         970   
SqFtTotBasement                           670                           0   
SqFtFinBasement                           390                           0   
FinBasementGrade                            6                           0   
SqFtGarageBasement                          0                           0   
SqFtGarageAttached                          0                         380   
DaylightBasement                                                        N   
SqFtOpenPorch                             150                           0   
SqFtEnclosedPorch                           0                          80   
SqFtDeck                                    0                           0   
HeatSystem                                  5                           5   
HeatSource                                  2                           2   
BrickStone                                  0                           0   
ViewUtilization                                                         N   
Bedrooms                                    3                           3   
BathHalfCount                               0                           0   
Bath3qtrCount                               0                           1   
BathFullCount                               1                           1   
FpSingleStory                               1                           1   
FpMultiStory                                0                           0   
FpFreestanding                              0                           0   
FpAdditional                                0                           0   
YrBuilt                                  1965                        1969   
YrRenovated                                 0                           0   
PcntComplete                                0                           0   
Obsolescence                                0                           0   
PcntNetCondition                            0                           0   
Condition                                   3                           3   
AddnlCost                                   0                           0   

                                            2                           3  \
Major                                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [30]:
small_bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']].copy()

In [31]:
small_bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512114 entries, 0 to 512113
Data columns (total 4 columns):
Major            512114 non-null int64
Minor            512114 non-null int64
SqFtTotLiving    512114 non-null int64
ZipCode          468423 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [32]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.8+ MB


In [33]:
merged_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [34]:
small_sales_df['Major'] = pd.to_numeric(small_sales_df['Major'])

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [4]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [35]:
small_sales_df['Major'] = pd.to_numeric(small_sales_df['Major'], errors='coerce')

Did it work?

In [19]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 24 columns):
ExciseTaxNbr          int64
Major                 float64
Minor                 object
DocumentDate          object
SalePrice             int64
RecordingNbr          object
Volume                object
Page                  object
PlatNbr               object
PlatType              object
PlatLot               object
PlatBlock             object
SellerName            object
BuyerName             object
PropertyType          int64
PrincipalUse          int64
SaleInstrument        int64
AFForestLand          object
AFCurrentUseLand      object
AFNonProfitUse        object
AFHistoricProperty    object
SaleReason            int64
PropertyClass         int64
SaleWarning           object
dtypes: float64(1), int64(7), object(16)
memory usage: 371.0+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [36]:
small_sales_df['Minor'] = pd.to_numeric(small_sales_df['Minor'], errors='coerce')

In [37]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.8+ MB


Now, let's try our join again.

In [40]:
merged_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

In [41]:
merged_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [42]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1446234 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1446234 non-null float64
Minor            1446234 non-null float64
DocumentDate     1446234 non-null object
SalePrice        1446234 non-null int64
SqFtTotLiving    1446234 non-null int64
ZipCode          1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 77.2+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [43]:
merged_data.loc[merged_data['ZipCode'].isna()].head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91  334330.0  1343.0   05/30/2006          0           4600     NaN
92  334330.0  1343.0   05/30/2006          0           4600     NaN
93  334330.0  1343.0   11/26/2001          0           4600     NaN
94  334330.0  1343.0   05/30/2006          0           4600     NaN
95  334330.0  1343.0   06/30/2016          0           4600     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [44]:
merged_data = merged_data.loc[~merged_data['ZipCode'].isna(), :]

merged_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [48]:
merged_data.loc[merged_data.SalePrice < 0, 'SalePrice'] = np.nan


In [50]:
merged_data.loc[merged_data.SqFtTotLiving == 0, 'SqFtTotLiving'] = np.nan

In [52]:
merged_data = merged_data.dropna()

In [54]:
merged_data.SqFtTotLiving.sort_values()

902859         1.0
902861         1.0
472934         2.0
472935         2.0
472936         2.0
472933         2.0
472932         2.0
472931         2.0
1057109        2.0
1057108        2.0
1392257        2.0
1392259        2.0
1328479        2.0
81469          2.0
438517         2.0
438520         2.0
438518         2.0
438519         2.0
1033119       70.0
1033120       70.0
1033121       70.0
1033123       70.0
1033122       70.0
1427969      100.0
1427968      100.0
1441487      110.0
1040577      120.0
1171723      140.0
1171724      140.0
1171721      140.0
            ...   
664468     19040.0
664467     19040.0
664470     19040.0
1349825    19570.0
1349826    19570.0
281694     19840.0
281693     19840.0
281692     19840.0
281695     19840.0
281690     19840.0
281691     19840.0
863545     19970.0
1358167    20140.0
1358168    20140.0
1229850    20600.0
1061099    20760.0
1061097    20760.0
1320701    24130.0
1320704    24130.0
1320702    24130.0
1320703    24130.0
445062     2

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [62]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1329060 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1329060 non-null float64
Minor            1329060 non-null float64
DocumentDate     1329060 non-null object
SalePrice        1329060 non-null float64
SqFtTotLiving    1329060 non-null float64
ZipCode          1329060 non-null object
dtypes: float64(4), object(2)
memory usage: 71.0+ MB


In [76]:
merged_data['ZipCode'][:5]

0    98002
1    98002
2    98002
3    98002
4    98092
Name: ZipCode, dtype: object

In [97]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        x = str(x).strip()
        x = x[:5]
        _ = pd.to_numeric(x)
    except ValueError:
        print(x)
        return False
    return True

merged_data.loc[merged_data['ZipCode'].apply(is_integer) == False, 'ZipCode'] = np.nan
merged_data.dropna(inplace = True)

A
A
A
B
WA
WA
WA


In [98]:
merged_data.dtypes

Major            float64
Minor            float64
DocumentDate      object
SalePrice        float64
SqFtTotLiving    float64
ZipCode           object
dtype: object

### 3. Add a column for PricePerSqFt



In [104]:
merged_data['PricePerSqFt'] = merged_data.SalePrice / merged_data.SqFtTotLiving

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [101]:
merged_data[merged_data.DocumentDate.apply(lambda x: x[-4:]) == '2019']

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
379      785996.0   190.0   05/16/2019   808000.0         3000.0   98072
557      367890.0    30.0   03/13/2019   923000.0         2550.0   98144
646      214090.0   110.0   03/06/2019   390000.0         1780.0   98042
905      945920.0   125.0   01/29/2019  1750000.0         1400.0   98118
906      945920.0   125.0   04/12/2019        0.0         1400.0   98118
1084     121800.0   175.0   03/26/2019   556500.0         1460.0   98166
1121     922140.0   427.0   05/13/2019        0.0         1370.0   98115
1211     145360.0  2285.0   05/02/2019   690000.0         1910.0   98125
1246     350160.0   125.0   02/06/2019   935000.0         2460.0   98117
1345       4000.0   228.0   04/26/2019   369000.0         1560.0   98168
1360     334510.0   288.0   04/24/2019        0.0         1560.0   98056
1463     259772.0    70.0   03/22/2019   525000.0         2570.0   98058
1476      85900.0   115.0   03/21/2019   563750.0         1690.0   98106
1489      22603.0  9195.0   05/07/2019   700000.0         2200.0   98177
1683     856450.0   210.0   05/03/2019        0.0         1590.0   98034
1721      45200.0  1890.0   04/01/2019  1085000.0         1650.0   98107
1740     186240.0   105.0   03/19/2019   380000.0          800.0   98117
1895     721740.0   680.0   05/24/2019   780800.0          860.0   98122
2144     352106.0  9028.0   03/21/2019   595000.0         3580.0   98022
2319      30200.0   645.0   03/04/2019   380000.0         1750.0   98001
2651     784920.0   780.0   05/15/2019   405000.0         1410.0   98065
2911     152104.0  9136.0   01/24/2019   268000.0         1551.0   98001
2917     602150.0  3655.0   04/26/2019        0.0         1320.0   98117
2929     668990.0    20.0   04/30/2019   780000.0         2320.0   98155
3052      51850.0  1050.0   01/28/2019        0.0         2040.0   98056
3228     547850.0   110.0   03/13/2019        0.0         1200.0   98030
3238     117500.0   594.0   02/14/2019   785000.0         1540.0   98107
3510      98400.0   450.0   02/20/2019   409950.0         1850.0   98058
3514     413942.0   460.0   05/22/2019  1630000.0         3480.0   98006
3610     797320.0  2320.0   03/27/2019   540000.0         1240.0   98146
...           ...     ...          ...        ...            ...     ...
1441188  732756.0   190.0   05/17/2019   856980.0         3270.0   98045
1441220  873180.0   530.0   05/13/2019   382000.0         2000.0   98023
1441286  277660.0   116.0   03/18/2019   740000.0         1510.0   98117
1441288  357480.0   970.0   02/22/2019   595000.0         2300.0   98034
1441359  305270.0   704.0   04/03/2019   839950.0         1510.0   98117
1441705  110800.0   740.0   05/07/2019   308000.0         1400.0   98118
1441933  660022.0    40.0   04/16/2019   825000.0         2230.0   98074
1442083  726620.0    77.0   04/01/2019   740000.0         1530.0   98115
1442166  887900.0    39.0   01/14/2019   301000.0         1270.0   98032
1442379   42505.0  9025.0   05/20/2019     5865.0         1090.0   98033
1442638  255818.0    40.0   05/06/2019        0.0         1810.0   98038
1442879   92304.0  9521.0   01/17/2019   562500.0         2450.0   98168
1443105  138730.0    80.0   01/18/2019   510000.0         1390.0   98011
1443640  713790.0   840.0   05/07/2019   333000.0         1350.0   98092
1443776  142206.0  9083.0   03/25/2019   580000.0         2070.0   98038
1443791  732756.0   180.0   04/23/2019   773750.0         2830.0   98045
1443962   32306.0  9004.0   04/29/2019        0.0         1010.0   98027
1444311  712930.0  4285.0   02/08/2019   475000.0         2360.0   98118
1444392  408330.0  6300.0   04/25/2019        0.0         2760.0   98103
1444564  645342.0   310.0   04/22/2019   699900.0         3130.0   98038
1444578  541410.0   122.0   02/14/2019   485000.0         1050.0   98118
1444667  162040.0   180.0   02/07/2019  1099000.0         2660.0   98034
1445333  507400.0   125.0   04/08/2019     3275.0         1270.0   981

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [107]:
def in_seattle(x):
    seattle = ['98101', '98102', '98103', '98104', '98105', '98106', '98107', 
           '98108', '98109', '98112', '98115', '98116', '98117', '98118', 
           '98119', '98121', '98122', '98125', '98126', '98133', '98134', 
           '98136', '98144', '98146', '98154', '98164', '98174', '98177',
           '98178', '98195', '98199']
    return x in seattle
seattle_homes = merged_data[merged_data.ZipCode.apply(in_seattle)]

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [109]:
seattle_homes.PricePerSqFt.mean()

138.43337603667965